### DISCUSSION TITLE
Support for relational data in content collections

### DISCUSSION THREAD

**Discussion by tony-sull** [withastro/roadmap#525] [Proposal]
**Update:** this proposal has been moved to stage 2. [Continue discussion here](https://github.com/withastro/roadmap/issues/530)

### Body

# Summary

Content collections unlocks some huge benefits, especially the DX benefits of type-checked markdown frontmatter.  Adding support for relational content, i.e. a `blog` collection that references an `authors` collection, would be a huge step to allow more flexible content modeling

# Background & Motivation

Taking a blog post as the example, there will likely be author information thats reused across multiple blog posts.  To standardize updates when I'm changing something like an author's avatar image, it'd be nice to have an `authors` or `people` collection that each blog post can reference by slug.

Astro's content APIs could use that for type-checking to make sure a known author slug is used and prefetch the data so `blog.data.author` would be the actual data rather than a string that needs to be passed back into `getEntryBySlug`

# Goals

**Minimum goal** - type-safe references by slug. i.e. The blog collection would validate that `blog.data.author` is a valid slug, but wouldn't transform or preload that data at all

**Bonus points** - related content is pre-fetched and the slug is replaced with the related data. i.e.  once loaded, `blog.data.author` would be an object with the author's parsed content instead of the author's slug

# Example

[Hacky userland experiment](https://github.com/tony-sull/tonysull.co/blob/12f7df437d3886c97ade55ba13509d6e8d56059e/src/content/config.ts#L6)

```md
---
title: My blog post
author: tony
---

## Insert awesome post here
```

```md
---
name: Tony
avatar: uploads/avatar.png
---
```

```ts
import { image, defineCollection, relation, z } from 'astro:content'

export const collections = {
  blog: {
    schema: z.object({
      title: z.string(),
      author: relation("authors").default("tony")
    })
  },
  authors: {
    schema: z.object({
      name: z.string(),
      avatar: image()
    })
  }
}
```

**Comment by ElianCodes** [comment]
This would be very cool! 🔥🔥
